### ISSUE TITLE
The official may have updated the rules again. Now the request to https://gae2-spclient.spotify.com/playplay/v1/key/file_id will result in a 403 response.

### ISSUE THREAD

**Issue by xyb921866260** [es3n1n/re-unplayplay#12]

**Comment by es3n1n** [comment]
> I am sure this is because the token got banned again. They will have to update it manually and you will need to upgrade to make it work again. 

Correct.

> This is why hardcoding the tokens is not a good idea. 

Regarding hardcoded tokens - while not ideal, it's currently the most practical approach for this use case. I'm open to alternative implementation suggestions, however please remember that decryption routines and keys are getting generated randomly for each build so for automatic retrievals/proceeding you'll need some sort of IR that you'll symbolically evaluate or something of that sort.

> But there is very little transparency on how the playplay tokens are retrieved, thus making it unreasonably hard to retrieve the tokens by yourself.

As for token retrieval - while it does require reverse engineering skills, the process is documented [here](https://git.gay/uhwot/unplayplay/issues/3#issuecomment-2318). I'm working on updating both the token and decryption routines, and should have a fix ready in *hopefully* a few hours.

Thanks for your patience - as this is a hobby project I maintain in my free time, updates may sometimes take some time.

**Comment by es3n1n** [comment]
Fixed in v1.2.6

**Comment by xyloflake** [comment]
> > But there is very little transparency on how the playplay tokens are retrieved, thus making it unreasonably hard to retrieve the tokens by yourself.
> 
> As for token retrieval - while it does require reverse engineering skills, the process is documented [here](https://git.gay/uhwot/unplayplay/issues/3#issuecomment-2318).

That's again, essentially me haha. Let's just say I did spend about 14 hours trying to figure it out (this is my first time trying out reverse engineering) and eventually gave up because I couldn't find any decrypt main function nor a token and a playIntent key. Searching was quite troublesome. I tried a few methods. Went through all of the ascii string but didn't find anything. It would definitely take some exceptional skills. I contacted one of my friends with reverse engineering skills but he met the same fate.

If I could somehow know exactly how to do it, I would maybe try to automate it? 

Suischan in a librespot java thread mentions that no matter how many tests they ran, the token was always hardcoded. So maybe if there is a definite set of instructions to the token retrieval, then maybe it could be automated. Though I'm really skeptical about spotify coming after the automation because that seriously would do them considerable damage(?).

Thanks for giving this project your valuable time @es3n1n ! It means a lot :) (thanks to @suischan, uhwot and @glomatico as well).

**Comment by es3n1n** [comment]
@xyloflake what version of the app did you check exactly? A few months ago spotify added more stuff to the android app's playplay which does make it harder to recreate. If you're interested in recreating stuff in an easier way I would recommend you looking at older versions of application.

**Comment by xyloflake** [comment]
> @xyloflake what version of the app did you check exactly? A few months ago spotify added more stuff to the android app's playplay which does make it harder to recreate. If you're interested in recreating stuff in an easier way I would recommend you looking at older versions of application.

That might explain it! Thank you so much @es3n1n ! I will try again with some time on my hands but I was trying out the latest release as of last week 😅

**Comment by es3n1n** [comment]
> I was trying out the latest release as of last week 😅

Yeah their latest releases are slightly more complex to reverse engineer

**Comment by xyloflake** [comment]
> > I was trying out the latest release as of last week 😅
> 
> Yeah their latest releases are slightly more complex to reverse engineer

Thank you for the valuable information!

**Comment by SuisChan** [comment]
> I will try again with some time on my hands but I was trying out the latest release as of last week 😅

Try below 8.9.80 it should be fine
